packages/nouns-contracts/contracts/NounsAuctionHouseV2.sol

```diff
@@ -0,0 +1,294 @@
+    }
+
+    /**
+     * @notice Create an auction.
+     * @dev Store the auction details in the `auction` state variable and emit an AuctionCreated event.
+     * If the mint reverts, the minter was updated without pausing this contract first. To remedy this,
+     * catch the revert and pause this contract.
+     */
+    function _createAuction() internal {
+        try nouns.mint() returns (uint256 nounId) {
+            uint256 startTime = block.timestamp;
+            uint256 endTime = startTime + duration;
+
+            auction = Auction({
+                nounId: nounId,
+                amount: 0,
+                startTime: startTime,
+                endTime: endTime,
+                bidder: payable(0),
+                settled: false
+            });
+
+            emit AuctionCreated(nounId, startTime, endTime);
+        } catch Error(string memory) {
+            _pause();
+        }
+    }
+
+    /**
+     * @notice Settle an auction, finalizing the bid and paying out to the owner.
+     * @dev If there are no bids, the Noun is burned.
+     */
+    function _settleAuction() internal {
+        INounsAuctionHouse.Auction memory _auction = auction;
+
+        require(_auction.startTime != 0, "Auction hasn't begun");
+        require(!_auction.settled, 'Auction has already been settled');
+        require(block.timestamp >= _auction.endTime, "Auction hasn't completed");
+
+        auction.settled = true;
+
+        if (_auction.bidder == address(0)) {
+            nouns.burn(_auction.nounId);
+        } else {
+            nouns.transferFrom(address(this), _auction.bidder, _auction.nounId);
+        }
+
+        if (_auction.amount > 0) {
+            _safeTransferETHWithFallback(owner(), _auction.amount);
+        }
+
+        oracle.write(
+            uint32(block.timestamp),
+            uint16(_auction.nounId),
+            Noracle.ethPriceToUint48(_auction.amount),
+            _auction.bidder
+        );
+
+        emit AuctionSettled(_auction.nounId, _auction.bidder, _auction.amount);
+    }
+
+    /**
+     * @notice Transfer ETH. If the ETH transfer fails, wrap the ETH and try send it as WETH.
+     */
+    function _safeTransferETHWithFallback(address to, uint256 amount) internal {
+        if (!_safeTransferETH(to, amount)) {
+            IWETH(weth).deposit{ value: amount }();
+            IERC20(weth).transfer(to, amount);
+        }
+    }
+
+    /**
+     * @notice Transfer ETH and return the success status.
+     * @dev This function only forwards 30,000 gas to the callee.
+     */
+    function _safeTransferETH(address to, uint256 value) internal returns (bool) {
+        (bool success, ) = to.call{ value: value, gas: 30_000 }(new bytes(0));
+        return success;
+    }
+
+    function initializeOracle() external onlyOwner {
+        oracle.initialize();
+    }
+
+    function growPriceHistory(uint32 newCardinality) external {
+        uint32 current = oracle.cardinalityNext;
```

Comment:
gas: `oracle.cardinalityNext` is read again in `oracle.grow`
maybe oracle.grow can return both current and next to avoid the extra sload?